plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/expression/ExpressionConverter.java

```diff
@@ -0,0 +1,260 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hudi.expression;
+
+import com.google.common.collect.ImmutableList;
+import io.airlift.log.Logger;
+import io.airlift.slice.Slice;
+import io.trino.spi.TrinoException;
+import io.trino.spi.expression.Call;
+import io.trino.spi.expression.ConnectorExpression;
+import io.trino.spi.expression.Constant;
+import io.trino.spi.expression.FunctionName;
+import io.trino.spi.expression.StandardFunctions;
+import io.trino.spi.expression.Variable;
+import org.apache.hudi.expression.Expression;
+import org.apache.hudi.expression.Literal;
+import org.apache.hudi.expression.NameReference;
+import org.apache.hudi.expression.Predicate;
+import org.apache.hudi.expression.Predicates;
+import org.apache.hudi.internal.schema.Type;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static io.trino.spi.StandardErrorCode.INVALID_ARGUMENTS;
+import static java.util.Objects.requireNonNull;
+
+public class ExpressionConverter
+{
+    private static final Logger log = Logger.get(ExpressionConverter.class);
+
+    private static final List<String> CANDIDATE_FUNCTIONS_FOR_EI = ImmutableList.of(
+            HudiTrinoYearFunctionExpression.TRINO_FN_NAME,
+            HudiTrinoMonthFunctionExpression.TRINO_FN_NAME,
+            HudiTrinoDayFunctionExpression.TRINO_FN_NAME,
+            HudiTrinoHourFunctionExpression.TRINO_FN_NAME,
+            HudiTrinoDateFormatFunctionExpression.TRINO_FN_NAME,
+            HudiTrinoFromUnixFunctionExpression.TRINO_FN_NAME,
+            HudiTrinoSubstringFunctionExpression.TRINO_FN_NAME);
+
+    private final Set<Predicate> candidateCollector;
+
+    public ExpressionConverter(ConnectorExpression connectorExpression)
+    {
+        requireNonNull(connectorExpression, "connectorExpression is null");
+        this.candidateCollector = new HashSet<>();
```

Comment:
Does this collection only contain one or no entry based on the logic right now?

Reply:
`tryAddAsCandidate` will attempt to add candidates. It will contain contain at least one or no entry based on the logic.